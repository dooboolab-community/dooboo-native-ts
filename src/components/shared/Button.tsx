import React, { Component } from 'react';
import styled from 'styled-components/native';
import {
  ActivityIndicator,
  TouchableOpacity,
  ViewStyle,
  TextStyle,
  ImageStyle,
} from 'react-native';

import { ratio, colors } from '../../utils/Styles';

const StyledButton = styled.View`
  background-color: transparent;
  align-self: center;
  border-radius: 4;
  border-width: 2;
  width: 320;
  height: 52;
  border-color: white;

  align-items: center;
  justify-content: center;
`;

const StyledButtonDisabled = styled(StyledButton)`
  background-color: rgb(224,224,224);
  border-color: rgb(200,200,200);
`;

const StyledText = styled.Text`
  font-size: 14;
  color: white;
`;

const StyledTextDisabled = styled(StyledText)`
  color: #969696;
`;

const StyledImage = styled.Image`
  width: 24;
  height: 24;
  position: absolute;
  left: 16;
`;

interface IProps {
  testID?: string;
  isLoading?: boolean;
  isDisabled?: boolean;
  onClick?: () => void;
  style?: ViewStyle;
  disabledStyle?: ViewStyle;
  textStyle?: TextStyle;
  imgLeftSrc?: any;
  imgLeftStyle?: ImageStyle;
  indicatorColor?: string;
  activeOpacity?: number;
  text?: string;
}

function Button(props: IProps) {
  if (props.isDisabled) {
    return (
      <StyledButtonDisabled style={props.disabledStyle}>
        <StyledTextDisabled style={props.textStyle}>{props.text}</StyledTextDisabled>
      </StyledButtonDisabled>
    );
  }
  if (props.isLoading) {
    return (
      <StyledButton style={props.style}>
        <ActivityIndicator size='small' color={props.indicatorColor} />
      </StyledButton>
    );
  }
  return (
    <TouchableOpacity
      activeOpacity={props.activeOpacity}
      onPress={props.onClick}
    >
      <StyledButton style={props.style}>
        {
          props.imgLeftSrc
            ? <StyledImage
              style={props.imgLeftStyle}
              source={props.imgLeftSrc}
            />
            : null
        }
        <StyledText style={props.textStyle}>{props.text}</StyledText>
      </StyledButton>
    </TouchableOpacity>
  );
}

Button.defaultProps = {
  isLoading: false,
  isDisabled: false,
  indicatorColor: 'white',
  activeOpacity: 0.5,
};

export default Button;
