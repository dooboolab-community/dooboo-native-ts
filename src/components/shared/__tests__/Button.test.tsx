import 'react-native';

import * as React from 'react';

import { RenderResult, act, fireEvent, render } from '@testing-library/react-native';

import Button from '../Button';
import { ThemeType } from '@dooboo-ui/native-theme';
import { createTestElement } from '../../../../test/testUtils';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: React.ReactElement;
let testingLib: RenderResult;

describe('[Button]', () => {
  let cnt = 1;

  beforeEach(() => {
    props = {
      onClick: (): number => cnt++,
      testID: 'btn',
    };
    component = createTestElement(<Button {...props} />);
  });

  it('[ThemeType.Light] renders without crashing', () => {
    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  it('should render [ThemeType.Dark] without crashing', () => {
    component = createTestElement(<Button {...props} />, ThemeType.DARK);

    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  it('should render [isDisabled] status without crashing', () => {
    props.isDisabled = true;
    component = createTestElement(<Button {...props} />);

    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  it('should render [isLoading] status without crashing', () => {
    props.isLoading = true;
    component = createTestElement(<Button {...props} />);

    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('interactions', () => {
    beforeEach(() => {
      testingLib = render(component);
    });

    it('should simulate onClick', () => {
      const btn = testingLib.queryByTestId('btn');
      act(() => {
        fireEvent.press(btn);
        fireEvent.press(btn);
      });
      expect(cnt).toBe(3);
    });
  });
});
