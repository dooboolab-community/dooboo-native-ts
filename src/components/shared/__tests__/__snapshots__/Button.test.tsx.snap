// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Button] [ThemeType.Light] renders without crashing 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  onPress={[Function]}
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#069ccd",
          "borderColor": "#069ccd",
          "borderRadius": 4,
          "borderWidth": 2,
          "height": 52,
          "justifyContent": "center",
          "width": 320,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 14,
          },
        ]
      }
    />
  </View>
</TouchableOpacity>
`;

exports[`[Button] should render [ThemeType.Dark] without crashing 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  onPress={[Function]}
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#069ccd",
          "borderColor": "#069ccd",
          "borderRadius": 4,
          "borderWidth": 2,
          "height": 52,
          "justifyContent": "center",
          "width": 320,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 14,
          },
        ]
      }
    />
  </View>
</TouchableOpacity>
`;

exports[`[Button] should render [isDisabled] status without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "rgb(224,224,224)",
        "borderColor": "rgb(200, 200, 200)",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 52,
        "justifyContent": "center",
        "width": 320,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#969696",
          "fontSize": 14,
        },
      ]
    }
  />
</View>
`;

exports[`[Button] should render [isLoading] status without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#069ccd",
        "borderColor": "#069ccd",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 52,
        "justifyContent": "center",
        "width": 320,
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="white"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;
